### PR TITLE
Local extra-data

### PR DESCRIPTION
**Pull request by manuq**

### CODE REVIEW

File: common/flatpak-dir.c [hunk 1]
```diff
@@ -2127,10 +2131,23 @@ flatpak_dir_pull_extra_data (FlatpakDir          *self,
 
       /* TODO: Download to disk to support resumed downloads on error */
 
-      ensure_soup_session (self);
-      bytes = flatpak_load_http_uri (self->soup_session, extra_data_uri, NULL, NULL,
-                                     extra_data_progress_report, &extra_data_progress,
-                                     cancellable, error);
+      home = g_file_new_for_path (g_get_home_dir ());
+      extra_local_filename = g_strconcat (".local/share/flatpak/extra-data/", extra_data_name, NULL);
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: Use flatpak_get_user_base_dir_location () instead

File: common/flatpak-dir.c [hunk 2]
```diff
@@ -2127,10 +2131,23 @@ flatpak_dir_pull_extra_data (FlatpakDir          *self,
 
       /* TODO: Download to disk to support resumed downloads on error */
 
-      ensure_soup_session (self);
-      bytes = flatpak_load_http_uri (self->soup_session, extra_data_uri, NULL, NULL,
-                                     extra_data_progress_report, &extra_data_progress,
-                                     cancellable, error);
+      base_dir = flatpak_get_user_base_dir_location ();
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: You should not free this return value.

File: common/flatpak-dir.c [hunk 3]
```diff
@@ -2127,10 +2131,23 @@ flatpak_dir_pull_extra_data (FlatpakDir          *self,
...
+      extra_local_dir = g_file_get_child (base_dir, "extra-data");
+      extra_local_file = g_file_get_child (g_file_get_child (extra_local_dir, extra_data_sha256), extra_data_name);
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: This leaks the intermediate child.
Use `flatpak_build_file (base_dir, "extra-data", extra_data_sha256, extra_data_name, NULL)` instead.

File: common/flatpak-dir.c [hunk 4]
```diff
@@ -2127,10 +2131,23 @@ flatpak_dir_pull_extra_data (FlatpakDir          *self,
...
+      if (g_file_query_exists (extra_local_file, cancellable))
+        {
+          g_debug ("Loading extra-data from local file %s", g_file_get_path (extra_local_file));
+          extra_local_stream = (GInputStream *) g_file_read (extra_local_file, cancellable, error);
+          bytes = g_input_stream_read_bytes (extra_local_stream, download_size, cancellable, error);
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: g_input_stream_read_bytes is just a more bindable version of g_input_stream_read(). It doesn't guarantee that you read the entire file, etc. Use g_file_load_contents instead and put the result (if the length is correct) in a GBytes.

**Comment by alexlarsson** [comment]
Also, can you drop the merge commits and squash this to a single commit.

**Comment by manuq** [comment]
Thanks I pushed a squashed commit with new changes.

For base_dir "You should not free this return value.", I interpreted that as "move the call outside the for loop". If you didn't mean that, please tell me and I'll change it again. Sorry I'm a glib newbie.

File: common/flatpak-dir.c [hunk 5]
```diff
@@ -2127,10 +2131,30 @@ flatpak_dir_pull_extra_data (FlatpakDir          *self,
...
+          gsize extra_local_size;
+          g_autofree char *extra_local_contents = NULL;
+          if (g_file_load_contents (extra_local_file, cancellable, &extra_local_contents, &extra_local_size, NULL, error) &&
+              extra_local_size == download_size)
+            {
+              bytes = g_bytes_new (extra_local_contents, extra_local_size);
+            }
+          else
+            {
+              return flatpak_fail (error, _("Wrong size for extra-data %s"), g_file_get_path (extra_local_file));
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: This will double-set error if it was set by g_file_load_contents, which will g_error

**Comment by alexlarsson** [comment]
Oh, sorry, i thought it returned a value that was not ref:ed, but i was wrong.

**Comment by manuq** [comment]
@alexlarsson , any other review?

**Pull request merged**